rfd/0229-agent-inventory-ui.md

```diff
@@ -0,0 +1,250 @@
+#### Web Endpoint
+
+A new `/webapi/sites/:site/instances` will be used by the Web UI to fetch a paginated list of instances, this endpoint will support
+various query parameters for pagination and filtering, including:
+
+- `search`: Filter by hostname.
+- `query`: Filter by a predicate language expression.
+- `services`: Filter to only return agents running the specified service(s).
+- `upgraders`: Filter to only return agents using the specified external upgrader(s).
+- `updaterGroup`: Filter to only return agents in the specified updater group.
+- `type`: Filters to return only agents, bot instances, or both. Default/undefined is both.
+
+The response will be a JSON object containing the page of agents requested.
+
+##### Example response:
+
+```json
+{
+  "instances": [
+    {
+      "id": "1",
+      "type": "agent",
+      "agent": {
+        "name": "server1",
+        "version": "16.1.3",
+        "services": ["ssh", "db", "desktop"],
+        "upgrader": {
+          "type": "unit",
+          "version": "v2",
+          "group": "group1"
+        }
+      }
+    },
+    {
+      "id": "2",
+      "type": "agent",
+      "agent": {
+        "name": "kube1",
+        "version": "16.5.4",
+        "services": ["kube"],
+        "upgrader": {
+          "type": "kube",
+          "version": "v2",
+          "group": "group1"
+        }
+      }
+    },
+    {
+      "id": "3",
+      "type": "bot",
+      "bot": {
+        "name": "bot1",
+        "version": "16.1.3",
+        "service": "kubernetes",
+        "health": "Unknown"
+      }
+    },
+    {
+      "id": "4",
+      "type": "agent",
+      "agent": {
+        "name": "agent1",
+        "version": "16.1.3",
+        "services": ["kube"],
+        "upgrader": {
+          "type": "kube",
+          "version": "v2",
+          "group": "group2"
+        }
+      }
+    }
+  ],
+  "totalCount": 4,
+  "startKey": ""
+}
+```
+
+#### RPC Service and Proto messages
+
+```protobuf
+service InventoryService {
+  // ListInstances returns a page of Agents and Bot Instances.
+  rpc ListInstances(ListInstancesRequest) returns (ListInstancesResponse);
+}
+
+// ListInstancesRequest is the request for listing instances.
+message ListInstancesRequest {
+  // page_size is the size of the page to return.
+  int32 page_size = 1;
+  // page_token is the next_page_token value returned from a previous ListInstances request, if any.
+  string page_token = 2;
+  // filters specify optional search criteria to limit which instances should be returned.
+  ListInstancesFilter filter = 3;
+}
+
+// ListInstancesResponse is the response from listing instances.
+message ListInstancesResponse {
+  // instances is the instances returned.
+  repeated Instance instances = 1;
+  // next_page_token contains the next page token to use as the start key for the next page of instances.
+  string next_page_token = 2;
+  // total_count is the total number of agents and bot instances in the cache.
+  int32 total_count = 3;
```

Comment:
Comments should call out the fact that including a total count is only sane because this API refuses requests when the cache is unhealthy.  We do not want people looking at this and thinking it's a pattern they should copy.  99% of the time, including a total count is a *very very* bad idea from a scaling perspective.

Reply:
Specifically, if we ever want to apply RBAC filtering to instances or bot instances we are going to be forced to walk through the whole list on every request, just to calculate the total count.

Reply:
You're absolutely right.  I hadn't considered RBAC filtering.  We should probably omit `total_count` entirely.